### ISSUE TITLE
In 3.3.3 (and before) compileExpression fails without add-opens on Java 17

### ISSUE THREAD

**Issue by petergeneric** [orphan-oss/ognl#160]
When Ognl.compileExpression calls ExpressionCompiler.compileExpression on Java 17 with 3.3.3 which uses the latest javassist, compile fails unless add-opens args have been added to the Java commandline.

The exception arises from this line in ExpressionCompiler:

```
Class clazz = pool.toClass(newClass);
```

I think this can be avoided - Javassist has an overload for `ClassPool.toClass` that lets you define a Neighbour Class, at which point it will use the new Java MethodHandles approach to define the class rather than the legacy ClassLoader.defineClass (which requires visibility changes, which is what fails without add-opens). If I instead change that line in ExpressionCompiler to:
```
Class clazz = pool.toClass(newClass, OgnlContext.class);
```

Then the expression compiles and works perfectly under Java 17.

I don't know for sure that it's safe to use OgnlContext.class as a general purpose neighbour class (don't know if MethodHandles defineClass is different in some fundamental way to the ClassLoader defineClass, and therefore how it would interact with other  environments that make more extensive use of Modules), so could we have a way to opt-in to it, perhaps the ability to pass a neighbour class in through an Ognl.compileExpression overload (either a class or a flag that would have the internal code pass OgnlContext.class as the neighbour) -- or even some public static flag on ExpressionCompiler?

**Comment by lukaszlenart** [comment]
What about extending `OgnlExpressionCompiler` with 
```java
default Optional<Class<?>> getClassNeighbor(Class<?> clazz) {
    return Optional.empty();
}
```

and the use this in
```java
Optional<Class<?>> classNeighbor = getClassNeighbor(expression.getClass());
Class<?> clazz;
if (classNeighbor.isPresent()) {
    clazz = pool.toClass(newClass, classNeighbor.get());
} else {
    clazz = pool.toClass(newClass);
}
```

then you can provide your own version of `ExpressionCompiler`

**Comment by petergeneric** [comment]
My understanding from reading the javassist javadoc is that the neighbour needs to be in the same package as CtClass, so the choice of neighbour class is tightly coupled to the ExpressionCompiler's choice of package for the CtClass, so this might be too general a solution?

The javadoc on `toClass(Class)` says "Warning: This method should not be used in Java 11 or later. Use toClass(CtClass, Class)." so my statement above might have been too cautious.

An alternative would be to abstract the pool.toClass call away entirely, as `Class<?> generateClass(ClassPool pool, CtClass newClass, Node expression)`?

But that said, this change would certainly work

**Comment by petergeneric** [comment]
Hi @lukaszlenart any thoughts on this? It'd be great to get expressions compiling without add-opens

**Comment by petergeneric** [comment]
Thanks! Related to this, I noticed that compilation produces very specific casts; do you think there's any benefit in trying to loosen casts? e.g. if the method being called comes from an interface, casting to that interface instead of the impl at compile-time.

The main motivation is to be able to handle a compiled expression getting different collection impls behind the scenes (e.g. a singleton list vs an arraylist)

Would you be open to a PR around this?

**Comment by petergeneric** [comment]
I've implemented my suggested change to `ExpressionCompiler` in my own branch at https://github.com/petergeneric/ognl/commit/8a29801e8c9a55e13905ddd60e69bd8e6e2b0a6b. I'm planning to also implement that casting constraint loosening on that branch before raising a PR, but just wanted to give you a heads-up in case you'd rather split into two separate POs.

**Comment by lukaszlenart** [comment]
@petergeneric any progress with this?

**Comment by petergeneric** [comment]
@lukaszlenart Sorry I missed this message... I had been looking at the loosening but ended up abandoning the work because it looked like it would have required more changes than I had been hoping for

**Comment by petergeneric** [comment]
P.S. still pretty sure that [the separate, simpler change to allow compiling on JDK17+](https://github.com/petergeneric/ognl/commit/8a29801e8c9a55e13905ddd60e69bd8e6e2b0a6b) will work, it's just a question of whether you agree if a fixed impl using OgnlContext as the neighbour is the right move (I think it is, based on my reading of the docs)

**Comment by lukaszlenart** [comment]
Looks good to me, just notice that I pushed a new release today which turn `OgnlContext` into a class instead of subclass of `Map`

**Comment by petergeneric** [comment]
Yep, that was the reason for checking in with this issue again, actually. I've rebased my changes against that commit and am just submitting a PR now

**Comment by lukaszlenart** [comment]
Closed by #214
